mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp

```diff
@@ -152,3 +152,152 @@ LogicalResult GENX::MatrixCopyOp::verify() {
+    return failure();
+  if (parser.resolveOperands(inputsOperands, inputTypes, inputsOperandsLoc,
+                             result.operands))
+    return failure();
+
+  // Parse optional attributes.
+  if (parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  // Parse lambda.
+  SmallVector<OpAsmParser::Argument> regionArgs;
+  if (parser.parseArgumentList(regionArgs, OpAsmParser::Delimiter::Paren,
+                               /*allowType=*/true, /*allowAttrs=*/true))
+    return failure();
+
+  Region *body = result.addRegion();
+  if (parser.parseRegion(*body, regionArgs))
+    return failure();
+
+  // Parse result type.
+  Type resType;
+  if (parser.parseColon() || parser.parseCustomTypeWithFallback(resType))
+    return failure();
+
+  result.addTypes(resType);
+
+  return success();
+}
+
+void GENX::MatrixMapOp::print(OpAsmPrinter &p) {
+  p << " ";
+  p.printStrippedAttrOrType(getScopeAttr());
+  p.printNewline();
+
+  SmallVector<Value> inputs{this->getMat()};
+  SmallVector<Type> inputTypes{this->getMat().getType()};
+  for (Value input : this->getInputs()) {
+    inputs.push_back(input);
+    inputTypes.push_back(input.getType());
+  }
+
+  p << "ins(";
+  p.printOperands(inputs);
+  p << " : " << inputTypes << ")";
+
+  SmallVector<::llvm::StringRef, 2> elidedAttrs;
+  elidedAttrs.push_back("scope");
+  p.printOptionalAttrDict((*this)->getAttrs(), elidedAttrs);
+
+  p.printNewline();
+  p << "(";
+  Block *body = getBody();
+  llvm::interleaveComma(body->getArguments(), p,
+                        [&](auto arg) { p.printRegionArgument(arg); });
+  p << ") ";
+
+  p.printRegion(getMapper(), /*printEntryBlockArgs=*/false);
+
+  p << " : ";
+  auto resType = getRes().getType();
+  if (auto validType = resType.dyn_cast<Type>())
+    p.printStrippedAttrOrType(validType);
+  else
+    p << resType;
+}
+
+LogicalResult GENX::MatrixMapOp::verify() {
+  // The scope attribute must be 'Subgroup' currently.
+  if (getScope() != GENX::Scope::Subgroup)
+    return this->emitOpError("scope attribute must have value 'Subgroup'");
+
+  SmallVector<Value> inputs{getMat()};
+  for (Value input : getInputs())
+    inputs.push_back(input);
+
+  Block *bodyBlock = getBody();
+  auto blockArgs = bodyBlock->getArguments();
+
+  // Checks that the arity of the `mapper` region is equal to the matrix
+  // argument plus the number of variadic arguments.
+  if (inputs.size() != blockArgs.size())
+    return emitOpError() << "expects number of operands to match the arity of "
+                            "mapper, but got: "
+                         << inputs.size() << " and " << blockArgs.size();
+
+  // The first parameters of the mapper should match the matrix element type.
+  auto matType = cast<GENX::JointMatrixType>(inputs.front().getType());
+  auto argTypes = bodyBlock->getArgumentTypes();
+  if (matType.getElementType() != argTypes.front())
+    return emitOpError() << "expected element type of input "
```

Comment:
missing invalid test case for this?

Reply:
There is one:
```
func.func @joint_matrix_mad(%mat : !genx.jointmatrix<8x16xf32, RowMajor>, %val: f32) {
  // expected-error @+1 {{'genx.matrix.map' op expected element type of input 'f32' to match bbArg type 'i32'}}
  %r = genx.matrix.map <Subgroup> 
    ins(%mat, %val : !genx.jointmatrix<8x16xf32, RowMajor>, f32)
    (%elem: i32, %v: i32) {
       %add = arith.addf %elem, %v : i32
       genx.yield %add : i32
    } : !genx.jointmatrix<8x16xf32, RowMajor>  
  llvm.return
}
```
but is out of order w.r.t. the code. Probably a result of merging. I have reordered the tests.

Reply:
ic, the missing one should be the one at line 298 instead.